### ISSUE TITLE
ExP2活性関数とELUの挙動比較(再確認）

### ISSUE THREAD

**Issue by Mokafe** [yasukikudan/exponential-probability#3]
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf

# Define the ExP2 activation function
def exp2_activation(x):
    return (tf.sign(x) + tf.nn.relu(x)) * (1 - tf.exp(-tf.abs(x)))

# Define the ELU activation function
def elu_activation(x, alpha=1.0):
    return tf.where(x >= 0, x, alpha * (tf.exp(x) - 1))

# Function to calculate the derivatives of the activation functions
def calculate_derivatives(activation_func, x_values):
    with tf.GradientTape() as tape:
        tape.watch(x_values)
        y = activation_func(x_values)
    return tape.gradient(y, x_values)

# Generate a series of numbers (x values)
x = np.linspace(-3, 3, 300)
x_tensor = tf.constant(x, dtype=tf.float32)

# Apply the activation functions
y_exp2 = exp2_activation(x_tensor)
y_elu = elu_activation(x_tensor)

# Calculate derivatives
dy_exp2 = calculate_derivatives(exp2_activation, x_tensor)
dy_elu = calculate_derivatives(elu_activation, x_tensor)

# Convert tensors to numpy arrays for plotting
x = x_tensor.numpy()
y_exp2, y_elu = y_exp2.numpy(), y_elu.numpy()
dy_exp2, dy_elu = dy_exp2.numpy(), dy_elu.numpy()

# Plotting the activation functions and their derivatives
plt.figure(figsize=(14, 6))

# Plot the activation functions
plt.subplot(1, 2, 1)
plt.plot(x, y_exp2, label='ExP2 Activation', color='red')
plt.plot(x, y_elu, label='ELU Activation', color='blue')
plt.title('Activation Functions')
plt.xlabel('Input value (x)')
plt.ylabel('Activated value')
plt.axhline(0, color='black', linewidth=0.5)
plt.axvline(0, color='black', linewidth=0.5)
plt.grid(True, linestyle='--', linewidth=0.5)
plt.legend()

# Plot the derivatives of the activation functions
plt.subplot(1, 2, 2)
plt.plot(x, dy_exp2, label='d(ExP2)/dx', color='red')
plt.plot(x, dy_elu, label='d(ELU)/dx', color='blue')
plt.title('Derivatives of Activation Functions')
plt.xlabel('Input value (x)')
plt.ylabel('Derivative')
plt.axhline(0, color='black', linewidth=0.5)
plt.axvline(0, color='black', linewidth=0.5)
plt.grid(True, linestyle='--', linewidth=0.5)
plt.legend()

plt.tight_layout()
plt.show()

![EXP2-ELU](https://github.com/yasukikudan/exponential-probability/assets/140834843/7a437d14-af01-41ce-93b4-4c73c5b6e396)
